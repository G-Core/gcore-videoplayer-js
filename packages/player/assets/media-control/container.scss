[data-player] {
  --bottom-panel: 40px;
}
.container {
  .media-control-notransition {
    transition: none !important;
  }

  .player-poster .play-wrapper {
    opacity: 1;
  }

  &.crop-video {
    [data-html5-video] {
      object-fit: cover;
    }
  }

  .player-poster .circle-poster {
    top: 50%;
    margin-top: -60px;
    left: 50%;
    margin-left: -60px;
    position: absolute;
    width: 120px;
    height: 120px;
    border: 2px solid white;
    border-radius: 50%;
    filter: "progid:DXImageTransform.Microsoft.Alpha(Opacity=60)";
    filter: alpha(opacity=60);
    opacity: 1;
    box-shadow: 0 0 1px 0 white;

    svg {
      margin-left: 5px;
      width: 80px;
    }
  }

  .spinner-three-bounce[data-spinner] > div {
    background-color: #ff5700;
  }
}
