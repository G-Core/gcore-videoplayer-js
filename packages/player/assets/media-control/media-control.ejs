<div class="media-control-background" data-background></div>

<div class="media-control-layer gcore-skin-bg-color" data-controls>
    <% var renderBar = function(name) { %>
        <div class="bar-container" data-<%= name %>>
            <div class="bar-background" data-<%= name %>>
                <div class="bar-fill-1" data-<%= name %>></div>
                <div class="bar-fill-2 gcore-skin-main-color" data-<%= name %>></div>
                <div class="bar-hover" data-<%= name %>></div>
            </div>
            <div class="bar-scrubber" data-<%= name %>>
                <div class="bar-scrubber-icon gcore-skin-main-color" data-<%= name %>></div>
            </div>
        </div>
    <% }; %>
    <% var renderSegmentedBar = function(name, segments) {
        segments = segments || 10; %>
    <div class="bar-container" data-<%= name %>>
        <div class="bar-background" data-<%= name %>>
            <div class="bar-fill-1 gcore-skin-main-color" data-<%= name %>></div>
        </div>
        <div class="bar-scrubber" data-<%= name %>>
            <div class="bar-scrubber-icon gcore-skin-main-color" data-<%= name %>></div>
        </div>
    </div>
    <% }; %>
    <% var renderDrawer = function(name, renderContent) { %>
        <div class="drawer-container" data-<%= name %>>
            <div class="drawer-icon-container" data-<%= name %>>
                <div class="drawer-icon media-control-icon gcore-skin-button-color" data-<%= name %>></div>
                <span class="drawer-text" data-<%= name %>></span>
            </div>
            <% renderContent(name); %>
        </div>
    <% }; %>
    <% var renderIndicator = function(name) { %>
        <div class="media-control-indicator gcore-skin-text-color" data-<%= name %>></div>
    <% }; %>
    <% var renderQuality = function(name) { %>
        <div class="media-control-quality" data-<%= name %>></div>
    <% }; %>
    <% var renderNerd = function(name) { %>
        <div class="media-control-nerd" data-<%= name %>></div>
    <% }; %>
    <% var renderMultiCamera = function(name) { %>
        <div class="media-control-multicamera" data-<%= name %>></div>
    <% }; %>
    <% var renderVR = function(name) { %>
        <div class="media-control-vr" data-<%= name %>></div>
    <% }; %>
    <% var renderButton = function(name) { %>
        <button type="button" class="media-control-button media-control-icon gcore-skin-button-color" data-<%= name %>
            aria-label="<%= name %>"></button>
    <% }; %>

    <% var templates = {
        bar: renderBar,
        segmentedBar: renderSegmentedBar,
    };
    var render = function (settingsList) {
        settingsList.forEach(function (setting) {

            if (setting === "seekbar") {
                renderBar(setting);
            } else if (setting === "volume") {
                renderDrawer(setting, settings.volumeBarTemplate ? templates[settings.volumeBarTemplate] : function (name) {
                    return renderSegmentedBar(name);
                });
            } else if (setting === "duration" || setting === "position" || setting === "clips") {
                renderIndicator(setting);
            } else if (setting === "quality") {
                renderQuality(setting)
            } else if (setting === "nerd") {
                renderNerd(setting)
            } else if (setting === "multicamera") {
                renderMultiCamera(setting)
            } else if (setting === "vr") {
                renderVR(setting)
            } else if (["playpause", "playstop", "fullscreen"].includes(setting)) {
                renderButton(setting);
            }
        });
    }; %>
    <% if (settings.left && settings.left.length) { %>
        <div class="media-control-left-panel" data-media-control>
            <% render(settings.left); %>
        </div>
    <% } %>
    <% if (settings.default && settings.default.length) { %>
        <div class="media-control-center-panel" data-media-control>
            <% render(settings.default); %>
        </div>
    <% } %>
    <% if (settings.right && settings.right.length) { %>
        <div class="media-control-right-panel" data-media-control>
            <% render(settings.right); %>
        </div>
    <% } %>
</div>
