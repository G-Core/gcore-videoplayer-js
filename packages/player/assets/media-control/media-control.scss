@use "sass:meta";
@use "container";
@use "width270";
@use "width370";

:root {
  --font-size-media-controls: 14px;
  --font-size-media-controls-dropdown: 16px;
  --player-vod-color: #005aff;
  --player-dvr-color: #fff;
  --player-live-color: #ff0101;
  --player-seekbar-current-color: #ff5700;
  --player-seekbar-buffer-color: #fff;
}

.media-control-skin-1[data-media-control-skin-1] {
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 9999;
  pointer-events: none;
  font-family: Roboto, "Open Sans", Arial, sans-serif;
  font-size: var(--font-size-media-controls);

  &.dragging {
    pointer-events: auto;
    cursor: grabbing !important;
    cursor: url("closed-hand.cur"), move;

    * {
      cursor: grabbing !important;
      cursor: url("closed-hand.cur"), move;
    }
  }

  .media-control-icon {
    line-height: 0;
    letter-spacing: 0;
    speak: none;
    color: #fff;
    vertical-align: middle;
    text-align: left;
    transition: all 0.1s ease;
  }

  .media-control-icon:hover {
    color: white;
  }

  &.media-control-hide {
    .media-control-background[data-background] {
      opacity: 0;
    }

    .media-control-layer[data-controls] {
      bottom: -50px;

      .bar-scrubber[data-seekbar] {
        opacity: 0;
      }
    }
  }

  .media-control-layer[data-controls] {
    position: absolute;
    bottom: 0;
    width: 100%;
    padding: 0 12px 0 8px;
    height: var(--bottom-panel);
    vertical-align: middle;
    pointer-events: auto;
    transition: bottom 0.4s ease-out;
    display: flex;
    justify-content: space-between;

    .media-control-left-panel {
      display: flex;
      align-items: center;
      gap: 0.5rem;
      justify-content: flex-start;
    }

    .media-control-center-panel {
      height: 100%;
      text-align: center;
      line-height: var(--bottom-panel);
    }

    .media-control-right-panel {
      display: flex;
      align-items: center;
      gap: 1rem;
      justify-content: flex-end;
      @media (max-width: 420px) {
        gap: 0.5rem;
        overflow: hidden;
      }
    }

    button.media-control-button {
      background-color: transparent;
      border: 0;
      padding: 0;
      cursor: pointer;
      display: inline-block;
      height: 20px;
      width: 24px;

      svg {
        height: 20px;

        path {
          fill: white;
        }
      }

      &:focus {
        outline: none;
      }

      &[data-play] {
        float: left;
        height: 100%;
      }

      &[data-pause] {
        float: left;
        height: 100%;
      }

      &[data-stop] {
        float: left;
        height: 100%;
      }

      &[data-fullscreen] {
        order: 100;
        background-color: transparent;
        border: 0;
        height: 40px;
      }

      &[data-hd-indicator] {
        background-color: transparent;
        border: 0;
        cursor: default;
        display: none !important;
        float: right;
        height: 100%;
      }
    }

    .media-control-indicator {

      &[data-position],
      &[data-duration] {
        display: flex;
        color: white;
        cursor: default;
        line-height: var(--bottom-panel);
        position: relative;
      }

      &[data-position] {
        margin: 1px 0 0 7px;
      }

      &[data-duration] {
        &::before {
          content: "|";
          margin-right: 7px;
        }

        color: rgb(255 255 255);
        opacity: 0.5;
        margin-top: 1px;
        margin-right: 6px;
      }
    }

    .media-clip-container {
      display: none;

      .media-clip-text,
      .media-clip-point {
        text-overflow: ellipsis;
        white-space: nowrap;
        overflow: hidden;
        display: inline-block;
        float: left;
        color: white;
        cursor: default;
        line-height: var(--bottom-panel);
        position: relative;
      }

      .media-clip-point {
        margin-right: 6px;
      }

      .media-clip-text {
        max-width: calc(80% - 240px);
      }
    }

    .bar-container[data-seekbar] {
      position: absolute;
      top: -11px;
      left: 0;
      display: inline-block;
      vertical-align: middle;
      width: 100%;
      height: 20px;
      cursor: pointer;

      .bar-background[data-seekbar] {
        width: 100%;
        height: 3px;
        position: relative;
        top: 8px;
        background-color: #666;

        .bar-fill-1[data-seekbar] {
          position: absolute;
          top: 0;
          left: 0;
          width: 0;
          height: 100%;
          background-color: var(--player-seekbar-buffer-color);
          transition: all 0.1s ease-out;
        }

        .bar-fill-2[data-seekbar] {
          position: absolute;
          top: 0;
          left: 0;
          width: 0;
          height: 100%;
          background-color: var(--player-seekbar-current-color);
          transition: all 0.1s ease-out;
        }
      }

      &.seek-disabled {
        cursor: default;
        display: none;
      }
    }

    .bar-container[data-seekbar].seek-disabled+.bar-scrubber[data-seekbar] {
      cursor: default;
      display: none;
    }

    .bar-scrubber[data-seekbar] {
      position: absolute;
      transform: translateX(-50%);
      top: -11.5px;
      left: 0;
      width: 20px;
      height: 20px;
      opacity: 1;
      transition: all 0.1s ease-out;

      .bar-scrubber-icon[data-seekbar] {
        position: absolute;
        left: 4.5px;
        top: 4.5px;
        width: 11px;
        height: 11px;
        border-radius: 50%;
        background-color: white;
      }
    }

    .drawer-container[data-volume] {
      display: flex;
      justify-content: flex-start;
      height: var(--bottom-panel);
      cursor: pointer;
      box-sizing: border-box;

      .drawer-icon-container[data-volume] {
        bottom: 0;

        .drawer-icon[data-volume] {
          background-color: transparent;
          border: 0;
          box-sizing: content-box;
          height: var(--bottom-panel);
          width: 20px;

          svg {
            height: 20px;
            position: relative;
            top: 3px;
            margin-top: 7px;

            path {
              fill: white;
            }
          }

          &.muted svg {
            margin-left: 2px;
          }
        }
      }

      .bar-container[data-volume] {
        position: relative;
        margin-left: 10px;
        top: 8px;
        width: 80px;
        height: 23px;
        padding: 3px 0;
        transition: width 0.2s ease-out;

        .bar-background[data-volume] {
          height: 3px;
          border-radius: 100px;
          position: relative;
          top: 7px;
          background-color: #666;

          .bar-fill-1[data-volume] {
            position: absolute;
            top: 0;
            left: 0;
            width: 0;
            height: 100%;
            border-radius: 100px;
            background-color: white;
            transition: all 0.1s ease-out;
          }

          .bar-fill-2[data-volume] {
            position: absolute;
            top: 0;
            left: 0;
            width: 0;
            height: 100%;
            background-color: var(--player-vod-color);
            transition: all 0.1s ease-out;
          }
        }

        .bar-scrubber[data-volume] {
          position: absolute;
          transform: translateX(-50%);
          top: 3px;
          margin-left: 3px;
          width: 16px;
          height: 16px;
          opacity: 1;
          transition: all 0.1s ease-out;

          .bar-scrubber-icon[data-volume] {
            position: absolute;
            left: 3px;
            top: 5px;
            width: 7px;
            height: 7px;
            border-radius: 50%;
            background-color: white;
          }
        }

        .segmented-bar-element[data-volume] {
          float: left;
          width: 4px;
          padding-left: 2px;
          height: 12px;
          opacity: 0.5;
          box-shadow: inset 2px 0 0 white;
          transition: transform 0.2s ease-out;

          &.fill {
            box-shadow: inset 2px 0 0 #fff;
            opacity: 1;
          }

          &:nth-of-type(1) {
            padding-left: 0;
          }

          &:hover {
            transform: scaleY(1.5);
          }
        }
      }
    }
  }
}

/* TODO distribute between plugins' styles */
.media-control-skin-1[data-media-control-skin-1] {
  .media-control-cc {
    button[data-cc-button] {
      display: flex;
      justify-content: center;
      padding: 0;
      align-items: center;

      &:hover {
        color: white;
      }
    }

    ul { // TODO extract common styles
      background-color: rgb(74 74 74 / 60%);
      border: none;
      width: auto;
      border-radius: 4px;
      bottom: 52px;
      margin-left: -28px;

      li {
        font-size: 16px;
        text-align: center;
        white-space: nowrap;
        height: 30px;

        a {
          height: 30px;
          padding: 5px 10px;
          color: #fffffe;

          &:hover {
            background-color: rgb(0 0 0 / 40%);
          }
        }

        &.current a {
          background-color: rgb(0 0 0 / 40%);
        }

        &:first-child {
          a {
            border-bottom-left-radius: 4px;
            border-bottom-right-radius: 4px;
          }
        }

        &:last-child {
          a {
            border-top-left-radius: 4px;
            border-top-right-radius: 4px;
          }
        }
      }
    }
  }

  .seek-time[data-seek-time] {
    height: 26px;
    line-height: 26px;
    bottom: 52px;
    border-radius: 3px;
    background-color: rgb(74 74 74 / 70%);

    span {
      letter-spacing: 0.8px;
      font-size: 14px;
      font-family: Roboto, "Open Sans", Arial, sans-serif;
    }

    span[data-seek-time] {
      padding-left: 8px;
      padding-right: 8px;
    }

    span[data-duration] {
      display: none !important;
    }
  }
}