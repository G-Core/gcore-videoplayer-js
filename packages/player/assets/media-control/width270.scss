.media-control-skin-1.w270 {
  .media-control-layer[data-controls] {
    .drawer-container[data-volume] {
      .bar-container[data-volume] {
        transform: rotate(270deg);
        float: none;
        display: block;
        position: absolute;

        margin: 0;
        top: -40px;
        padding: 0;

        margin-left: -32px;
        margin-top: -3px;
        width: 80px;

        .bar-background-mask {
          position: absolute;
          width: 100%;
          height: 100%;
          left: -5px;
        }

        &.volume-bar-hide {
          display: none;
        }

        .bar-background {
          margin-left: 11px;
          top: 7px;
          width: 80px;
        }

        .bar-scrubber {
          margin-left: 11px;
          top: 1px;
        }
      }
    }
  }
  .media-control-dd__popup {
    max-width: 114px;
  }
}