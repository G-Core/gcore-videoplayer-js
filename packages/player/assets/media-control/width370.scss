.media-control-skin-1[data-media-control-skin-1].w370 {
  .media-control-cc {
    width: 28px;

    svg {
      height: 17px;
    }
  }

  .share_plugin[data-share] {
    .share-container {
      top: calc(50% - 150px);
      left: calc(50% - 125px);
      width: 250px;
      height: calc(100% - 32px);
      max-height: 300px;
      transform: none;

      .share-container-header {
        border: none;

        .share-container-header--title {
          visibility: hidden;
        }
      }
    }
  }

  .media-control-quality,
  .media-control-audiotracks {
    display: block;
  }

  .level-selector[data-level-selector],
  .media-control-audiotracks[data-audiotracks] {
    margin-top: 3px;
    margin-right: 10px;

    ul {
      bottom: 30px;
      width: 50px;

      li {
        height: 23px;
        font-size: 14px;

        a {
          height: 23px;
          padding: 2px 5px;
        }
      }
    }

    button[data-level-selector-button],
    button[data-audiotracks-button] {

      .quality-text,
      .audio-text {
        font-size: 13px;
      }

      span.quality-arrow,
      span.audio-arrow {
        width: 7px;
        height: 5px;
        margin-left: 4px;
        margin-top: 11px;
      }
    }
  }

  .media-control-multicamera {
    .multicamera {
      margin-top: 0;
      margin-right: 10px;

      button {
        height: 32px;

        svg {
          height: 17px;
          margin: 0;
        }
      }
    }
  }

  .media-control-layer[data-controls] {
    padding-left: 10px;
    padding-right: 12px;

    .media-control-indicator {
      line-height: 32px;

      &[data-position],
      &[data-duration] {
        font-size: 11px;
        line-height: 32px;
      }
    }

    .dvr-controls {
      height: 32px;

      .live-button {
        margin-left: 10px;
        height: 32px;
        font-size: 12px;
        line-height: 32px;
        text-shadow: none;
        letter-spacing: 0.6px;

        &::before {
          width: 8px;
          height: 8px;
          margin-right: 4px;
        }
      }

      .live-info {
        margin-left: 10px;
        height: 32px;
        font-size: 12px;
        line-height: 32px;
        text-shadow: none;
        letter-spacing: 0.6px;

        &::before {
          width: 8px;
          height: 8px;
          margin-right: 4px;
        }
      }

    }

    button.media-control-button {
      height: 32px;

      &[data-fullscreen] {
        height: 33px;
      }

      svg {
        height: 17px;
      }
    }

    .media-clip-container,
    .media-control-clips {

      .media-clip-text {
        line-height: 33px;
        font-size: 11px;
      }

      .media-clip-text {
        max-width: 50px;
      }
    }

    .drawer-container[data-volume] {
      height: 32px;
      margin-right: 8px;

      .drawer-icon-container[data-volume] {
        .drawer-icon[data-volume] {
          height: 32px;

          svg {
            height: 17px;
            margin-top: 5px;
          }
        }
      }

      .bar-container[data-volume] {
        display: none;

        &.volume-bar-hide {
          width: 0;
          height: 12px;
          top: 9px;
          padding: 0;
        }
      }
    }
  }
}