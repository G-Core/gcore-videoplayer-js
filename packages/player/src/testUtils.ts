import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
