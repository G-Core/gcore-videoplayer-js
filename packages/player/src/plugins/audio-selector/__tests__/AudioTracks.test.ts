import { beforeEach, describe, expect, it, vi } from 'vitest'
import { Events } from '@clappr/core'

import { ExtendedEvents } from '../../media-control/MediaControl'
import { AudioTracks } from '../AudioTracks'

import { createMockCore, createMockMediaControl } from '../../../testUtils'
// import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'

// Logger.enable('*')
// setTracer(new LogTracer('AudioTracks.test'))

const TRACKS = [
  { id: '1', label: 'English', language: 'en', track: {} },
  { id: '2', label: 'Spanish', language: 'es', track: {} },
]

describe('AudioTracks', () => {
  let core: any
  let mediaControl: any
  let audioTracks: AudioTracks
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    core.getPlugin = vi.fn().mockImplementation((name: string) => {
      if (name === 'media_control') return mediaControl
      return null
    })
    audioTracks = new AudioTracks(core)
    core.emit(Events.CORE_READY)
    core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
  })
  describe('basically', () => {
    it('should render button', () => {
      expect(audioTracks.$el.find('#gplayer-audiotracks-button').length).toBe(1)
    })
    it('should render menu hidden', () => {
      expect(audioTracks.$el.find('#gplayer-audiotracks-menu').css('display')).toBe('none')
    })
  })
  describe('until audio tracks are available', () => {
    it('should not mount', () => {
      expect(mediaControl.slot).not.toHaveBeenCalledWith(
        'audiotracks',
        expect.anything(),
      )
    })
  })
  describe('when audio tracks are available', () => {
    beforeEach(() => {
      emitTracksAvailable(core, TRACKS)
    })
    it('should render menu', () => {
      expect(audioTracks.el.innerHTML).toMatchSnapshot()
      expect(
        audioTracks.$el.find('#gplayer-audiotracks-menu').css('display'),
      ).toBe('none')
      const trackItems = audioTracks.$el.find('#gplayer-audiotracks-menu li')
      expect(trackItems.length).toBe(2)
      expect(trackItems.eq(0).text().trim()).toBe('English')
      expect(trackItems.eq(1).text().trim()).toBe('Spanish')
    })
    it('should mount', () => {
      expect(mediaControl.slot).toHaveBeenCalledWith(
        'audiotracks',
        audioTracks.$el,
      )
    })
    describe('when button is clicked', () => {
      beforeEach(() => {
        audioTracks.$el.find('#gplayer-audiotracks-button').click()
      })
      it('should show menu', () => {
        expect(audioTracks.$el.html()).toMatchSnapshot()
        expect(
          audioTracks.$el.find('#gplayer-audiotracks-menu').css('display'),
        ).not.toBe('none')
      })
      it('should collapse all other menus', () => {
        expect(mediaControl.trigger).toHaveBeenCalledWith(
          ExtendedEvents.MEDIACONTROL_MENU_COLLAPSE,
          'audio_tracks',
        )
      })
      describe('when audio track is selected', () => {
        beforeEach(() => {
          audioTracks.$el
            .find('#gplayer-audiotracks-menu [data-item="2"]')
            .click()
        })
        it('should switch to the selected audio track', () => {
          expect(core.activeContainer.switchAudioTrack).toHaveBeenCalledWith(
            '2',
          )
        })
        it('should hide the menu', () => {
          expect(audioTracks.$el.html()).toMatchSnapshot()
          expect(
            audioTracks.$el.find('#gplayer-audiotracks-menu').css('display'),
          ).toBe('none')
          expect(audioTracks.$el.find('#gplayer-audiotracks-button').attr('aria-expanded')).toBe('false')
        })
        it('should add changing class to the button', () => {
          expect(
            audioTracks.$el.find('#gplayer-audiotracks-button').hasClass('changing'),
          ).toBe(true)
        })
        describe('when current audio track changes', () => {
          beforeEach(() => {
            core.activePlayback.currentAudioTrack =
              core.activePlayback.audioTracks[1]
            core.activePlayback.emit(
              Events.PLAYBACK_AUDIO_CHANGED,
              core.activePlayback.currentAudioTrack,
            )
            core.activeContainer.emit(
              Events.CONTAINER_AUDIO_CHANGED,
              core.activePlayback.currentAudioTrack,
            )
          })
          it('should update button class', () => {
            expect(
              audioTracks.$el
                .find('#gplayer-audiotracks-button')
                .hasClass('changing'),
            ).toBe(false)
          })
          it('should update button label', () => {
            expect(
              audioTracks.$el
                .find('#gplayer-audiotracks-button')
                .text()
                .replace(/\/assets.*\.svg/g, '')
                .trim(),
            ).toBe('Spanish')
          })
          it('should highlight the selected menu item', () => {
            const selectedItem = audioTracks.$el.find(
              '#gplayer-audiotracks-menu .current',
            )
            expect(selectedItem.text().trim()).toBe('Spanish')
            expect(
              selectedItem
                .find('a[data-item]')
                .hasClass('gcore-skin-active'),
            ).toBe(true)
          })
          it('should unhighlight any previously highlighted menu item', () => {
            expect(
              audioTracks.$el.find('#gplayer-audiotracks-menu li.current').length,
            ).toBe(1)
            expect(
              audioTracks.$el.find(
                '#gplayer-audiotracks-menu a.gcore-skin-active[data-item]',
              ).length,
            ).toBe(1)
          })
        })
      })
    })
    describe('when button is clicked twice', () => {
      beforeEach(() => {
        audioTracks.$el.find('#gplayer-audiotracks-button').click()
        audioTracks.$el.find('#gplayer-audiotracks-button').click()
      })
      it('should collapse the menu', () => {
        expect(audioTracks.$el.find('#gplayer-audiotracks-menu').css('display')).toBe('none')
        expect(audioTracks.$el.find('#gplayer-audiotracks-button').attr('aria-expanded')).toBe('false')
      })
    })
    describe('when media control is rendered', () => {
      beforeEach(() => {
        mediaControl.slot.mockClear()
        mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
      })
      it('should remount', () => {
        expect(mediaControl.slot).toHaveBeenCalledWith(
          'audiotracks',
          audioTracks.$el,
        )
      })
    })
  })
  describe('when container is clicked', () => {
    beforeEach(() => {
      emitTracksAvailable(core, TRACKS)
      audioTracks.$el.find('#gplayer-audiotracks-button').click()
      core.activeContainer.emit(Events.CONTAINER_CLICK)
    })
    it('should hide the menu', () => {
      expect(audioTracks.$el.find('#gplayer-audiotracks-menu').css('display')).toBe('none')
      expect(audioTracks.$el.find('#gplayer-audiotracks-button').attr('aria-expanded')).toBe('false')
    })
  })
})

function emitTracksAvailable(core: any, tracks: any[]) {
  core.activePlayback.audioTracks = tracks
  core.activePlayback.currentAudioTrack = tracks[0]
  core.activePlayback.emit(
    Events.PLAYBACK_AUDIO_AVAILABLE,
    core.activePlayback.audioTracks,
  )
  core.activeContainer.emit(
    Events.CONTAINER_AUDIO_AVAILABLE,
    core.activePlayback.audioTracks,
  )
}
