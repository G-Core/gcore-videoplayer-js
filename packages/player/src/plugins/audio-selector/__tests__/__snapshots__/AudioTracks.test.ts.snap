// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AudioTracks > when audio tracks are available > should render menu 1`] = `
"<button class="gcore-skin-button-color media-control-dd" id="gplayer-audiotracks-button" aria-haspopup="menu" aria-expanded="false">
    <span class="media-control-dd__text" id="gplayer-audiotracks-button-text"></span>
    <span class="media-control-dd__arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu media-control-dd__popup" id="gplayer-audiotracks-menu" role="menu" style="display: none;">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioTracks > when audio tracks are available > when button is clicked > should show menu 1`] = `
"<button class="gcore-skin-button-color media-control-dd" id="gplayer-audiotracks-button" aria-haspopup="menu" aria-expanded="true">
    <span class="media-control-dd__text" id="gplayer-audiotracks-button-text"></span>
    <span class="media-control-dd__arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu media-control-dd__popup" id="gplayer-audiotracks-menu" role="menu">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;

exports[`AudioTracks > when audio tracks are available > when button is clicked > when audio track is selected > should hide the menu 1`] = `
"<button class="gcore-skin-button-color media-control-dd changing" id="gplayer-audiotracks-button" aria-haspopup="menu" aria-expanded="false">
    <span class="media-control-dd__text" id="gplayer-audiotracks-button-text"></span>
    <span class="media-control-dd__arrow">/assets/icons/old/quality-arrow.svg</span>
</button>
<ul class="gcore-skin-bg-color menu media-control-dd__popup" id="gplayer-audiotracks-menu" role="menu" style="display: none;">
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="1" role="menuitemradio" aria-checked="false">
                English
            </a>
        </li>
    
        <li class="">
            <a href="#" class="gcore-skin-text-color" data-item="2" role="menuitemradio" aria-checked="false">
                Spanish
            </a>
        </li>
    
</ul>
"
`;
